### PR TITLE
wire in DefaultCredentialsChain from aws_config as the credential provider for object_store

### LINKED ISSUE #912: Fix the behavior of auto-loading credentials chain on S3
# Problem Statement

Currently, Lance does not support a "default chain of credentials lookups", which makes the experience on AWS bad. 

AWS rust sdk has the desired functionalty
https://docs.rs/aws-config/latest/aws_config/default_provider/credentials/struct.DefaultCredentialsChain.html

Not sure how to integrate into `object_store` rust crate yet.

### PR DESCRIPTION
**Pull request by chebbyChefNEQ**
closes #912 

`object_store` has a custom S3 client which does not accept `CredentialProviders` from `aws_config`. This PR implements an adaptor that adapts a `DefaultCredentialsChain` to a `object_store::CredentialProvider`.

Screenshot 2023-06-26 at 6 54 13 PM

### CODE REVIEW

File: rust/src/io/object_store.rs [hunk 1]
```diff
@@ -52,16 +57,43 @@ impl std::fmt::Display for ObjectStore {
     }
 }
 
+#[derive(Debug)]
+struct AwsCredentialAdapter {
+    pub inner: DefaultCredentialsChain,
+}
+
+#[async_trait]
+impl CredentialProvider for AwsCredentialAdapter {
```
Comment (changhiskhan) [inline rust/src/io/object_store.rs:66]: Does this pickup credentials from EC2 IAM roles?

Ideally we'd support the following modes (forgot the right order):
1. Credentials as envvars
2. ~/.aws/credentials
3. IAM role on EC2 instance

Additionally, if the user has multiple profiles setup, then AWS_PROFILE envvar should also affect which sets of credentials are being used.

Comment (chebbyChefNEQ) [reply rust/src/io/object_store.rs:66]: I think that is the behavior for DefaultProviderChain, code here: https://github.com/awslabs/aws-sdk-rust/blob/d1962fd42d376c5c695877041d2a269fd1fa50cd/sdk/aws-config/src/default_provider/credentials.rs#L189-L193

The profile provider in that chain uses `AWS_PROFILE` envvar for selecting profile. (After many layers of abstractions, this is the code used for selecting profile name: https://github.com/awslabs/aws-sdk-rust/blob/d1962fd42d376c5c695877041d2a269fd1fa50cd/sdk/aws-config/src/profile/parser/source.rs#L59-L63)

Comment (chebbyChefNEQ) [reply rust/src/io/object_store.rs:66]: That said, let me test this.

File: rust/Cargo.toml [hunk 2]
```diff
@@ -37,7 +37,7 @@ async-trait = "0.1.60"
 byteorder = "1.4.3"
 chrono = "0.4.23"
 clap = { version = "4.1.1", features = ["derive"], optional = true }
-object_store = { version = "0.5.6", features = ["aws_profile", "gcp"] }
+object_store = { version = "0.6.0", features = ["aws", "gcp"] }
```
Comment (eddyxu) [inline rust/Cargo.toml:40]: do we need to make this sync with datafusion / arrow-rs?

Comment (chebbyChefNEQ) [reply rust/Cargo.toml:40]: Yeah you are right, bumping.

Comment (chebbyChefNEQ) [reply rust/Cargo.toml:40]: ok, we need object store `0.6.0` which is used by arrow `40.0`. BUT, datafusion `25.0` is still on `39.0`. going to put this PR on hold till the next datafusion release.

Note: there are some breaking changes. I'll make a separate PR when the new datafusion is released

**Review by eddyxu — APPROVED** [review]
It finally happens. Thanks @chebbyChefNEQ 

Please make sure it works locally and on EC2 w/ IAM agent.

**Pull request merged**